### PR TITLE
Hide irrelevant lines in suggestions to allow for suggestions that are far from each other to be shown

### PR DESCRIPTION
**Pull request by WaffleLapkin**
This is an attempt to fix suggestions one part of which is 6 lines or more far from the first. I've noticed "the problem" (of not showing some parts of the suggestion) here: https://github.com/rust-lang/rust/pull/97759#discussion_r889689230.

I'm not sure about the implementation (this big closure is just bad and makes already complicated code even more so), but I want to at least discuss the result.

Here is an example of how this changes the output:

Before:
```text
help: consider enclosing expression in a block
  |
3 ~     'l: { match () { () => break 'l,
4 |
5 |
6 |
7 |
8 |
...
```

After:
```text
help: consider enclosing expression in a block
  |
3 ~     'l: { match () { () => break 'l,
4 |
...
31|
32~ } };
  |
```

r? @estebank 
@rustbot label +A-diagnostics +A-suggestion-diagnostics

### CODE REVIEW

**Comment by estebank** [comment]
I'm ok with the results, need to look at the impl, particularly to check that it'll be graceful when handling more than 2 suggested changes in a single suggestion.

File: compiler/rustc_errors/src/emitter.rs [hunk 1]
```diff
@@ -1839,79 +1834,115 @@ impl EmitterWriter {
...
-                    // N + fn foo(bar: impl T) {
+            let mut unhighlighted_lines = Vec::new();
+            for (line_pos, (line, highlight_parts)) in lines.by_ref().zip(highlights).enumerate() {
+                debug!(%line_pos, %line, ?highlight_parts);
+
+                let print_line = |line_pos: usize,
+                                  line: &str,
+                                  highlight_parts: &Vec<SubstitutionHighlight>,
+                                  buffer: &mut StyledBuffer,
+                                  row_num: &mut usize| {
```
Comment (estebank) [inline compiler/rustc_errors/src/emitter.rs:1845]: Maybe moving this to its own method might help with readability.

Comment (WaffleLapkin) [reply compiler/rustc_errors/src/emitter.rs]: I tried, but I'm not sure if it's better, there are a lot of variables / state that needs to be passed around.

**Comment by estebank** [comment]
> @WaffleLapkin
[Add back MAX_SUGGESTION_HIGHLIGHT_LINES so clippy is happy & buildable](https://github.com/rust-lang/rust/pull/97798/commits/1961d9e57e2a5d668a1a7028f786147291ba325b)

We can modify clippy directly from this repo as well.

File: compiler/rustc_errors/src/emitter.rs [hunk 2]
```diff
@@ -1839,79 +1833,94 @@ impl EmitterWriter {
...
-                        }
-                        _ => {
-                            buffer.puts(row_num, max_line_num_len + 1, "~ ", Style::Addition);
-                        }
-                    }
-                } else {
-                    draw_col_separator(&mut buffer, row_num, max_line_num_len + 1);
+            let mut unhighlighted_lines = Vec::new();
+            for (line_pos, (line, highlight_parts)) in lines.by_ref().zip(highlights).enumerate() {
+                debug!(%line_pos, %line, ?highlight_parts);
```
[the remark this answers is not in the document]
Comment (WaffleLapkin) [reply compiler/rustc_errors/src/emitter.rs:1838]: It uses `Display` instead of `Debug` :D

https://docs.rs/tracing/latest/tracing/#recording-fields

**Comment by WaffleLapkin** [comment]
> We can modify clippy directly from this repo as well.

Oh, nice! Didn't know about that.

**Comment by JohnTitor** [comment]
Failed in rollup: https://github.com/rust-lang/rust/runs/6897040285?check_suite_focus=true
Seems we have to bless clippy tests, @bors r-

**Comment by WaffleLapkin** [comment]
I think it should be fine now

**Pull request merged**